File: src/gpt_from_scratch/eda/v2.cpp
#include <torch/torch.h>
#include <torch/nn/functional.h>
#include <iostream>
#include <fstream>
#include <vector>
#include <random>

// Hyperparameters
const int batch_size = 64;
const int block_size = 256;
const int max_iters = 5000;
const int eval_interval = 500;
const float learning_rate = 3e-4;
const int eval_iters = 200;
const int n_embd = 384;
const int n_head = 6;
const int n_layer = 6;
const float dropout = 0.2;

// Vocabulary
std::vector<char> chars;
int vocab_size;
std::unordered_map<char, int> stoi;
std::unordered_map<int, char> itos;

// Data
std::vector<int64_t> train_data;
std::vector<int64_t> val_data;

// Model
torch::Device device = torch::cuda::is_available() ? torch::kCUDA : torch::kCPU;

class HeadImpl : public torch::nn::Module {
public:
    HeadImpl(int head_size) {
        key = register_module("key", torch::nn::Linear(n_embd, head_size, false));
        query = register_module("query", torch::nn::Linear(n_embd, head_size, false));
        value = register_module("value", torch::nn::Linear(n_embd, head_size, false));
        tril = torch::tril(torch::ones(block_size, block_size));
        dropout = torch::nn::Dropout(dropout);
    }

    torch::Tensor forward(torch::Tensor x) {
        torch::Tensor k = key(x);
        torch::Tensor q = query(x);
        torch::Tensor wei = torch::matmul(q, k.transpose(-2, -1)) * std::pow(n_embd, -0.5);
        wei.masked_fill_(tril[0][0].unsqueeze(0).unsqueeze(0) == 0, -std::numeric_limits<float>::infinity());
        wei = torch::softmax(wei, -1);
        wei = dropout(wei);
        torch::Tensor v = value(x);
        torch::Tensor out = torch::matmul(wei, v);
        return out;
    }

    torch::nn::Linear key;
    torch::nn::Linear query;
    torch::nn::Linear value;
    torch::Tensor tril;
    torch::nn::Dropout dropout;
};
TORCH_MODULE(Head);

class MultiHeadAttentionImpl : public torch::nn::Module {
public:
    MultiHeadAttentionImpl(int num_heads, int head_size) {
        for (int i = 0; i < num_heads; ++i) {
            heads.push_back(register_module("head" + std::to_string(i), Head(head_size)));
        }
        proj = register_module("proj", torch::nn::Linear(n_embd, n_embd));
        dropout = torch::nn::Dropout(dropout);
    }

    torch::Tensor forward(torch::Tensor x) {
        std::vector<torch::Tensor> outputs;
        for (auto& head : heads) {
            outputs.push_back(head(x));
        }
        torch::Tensor out = torch::cat(outputs, -1);
        out = dropout(proj(out));
        return out;
    }

    std::vector<Head> heads;
    torch::nn::Linear proj;
    torch::nn::Dropout dropout;
};
TORCH_MODULE(MultiHeadAttention);

class FeedForwardImpl : public torch::nn::Module {
public:
    FeedForwardImpl() {
        net = torch::nn::Sequential(
            torch::nn::Linear(n_embd, 4 * n_embd),
            torch::nn::ReLU(),
            torch::nn::Linear(4 * n_embd, n_embd),
            torch::nn::Dropout(dropout)
        );
    }

    torch::Tensor forward(torch::Tensor x) {
        return net->forward(x);
    }

    torch::nn::Sequential net;
};
TORCH_MODULE(FeedForward);

class BlockImpl : public torch::nn::Module {
public:
    BlockImpl() {
        sa = register_module("sa", MultiHeadAttention(n_head, n_embd / n_head));
        ffwd = register_module("ffwd", FeedForward(n_embd));
        ln1 = register_module("ln1", torch::nn::LayerNorm(n_embd));
        ln2 = register_module("ln2", torch::nn::LayerNorm(n_embd));
    }

    torch::Tensor forward(torch::Tensor x) {
        x = x + sa->forward(ln1->forward(x));
        x = x + ffwd->forward(ln2->forward(x));
        return x;
    }

    MultiHeadAttention sa;
    FeedForward ffwd;
    torch::nn::LayerNorm ln1;
    torch::nn::LayerNorm ln2;
};
TORCH_MODULE(Block);

class BigramLanguageModelImpl : public torch::nn::Module {
public:
    BigramLanguageModelImpl() {
        token_embedding_table = register_module("token_embedding_table", torch::nn::Embedding(vocab_size, n_embd));
        position_embedding_table = register_module("position_embedding_table", torch::nn::Embedding(block_size, n_embd));
        for (int i = 0; i < n_layer; ++i) {
            blocks->push_back(register_module("block" + std::to_string(i), Block()));
        }
        ln_f = register_module("ln_f", torch::nn::LayerNorm(n_embd));
        lm_head = register_module("lm_head", torch::nn::Linear(n_embd, vocab_size));
    }

    torch::Tensor forward(torch::Tensor idx, torch::Tensor targets = nullptr) {
        int B = idx.size(0);
        int T = idx.size(1);

        torch::Tensor token_emb = token_embedding_table->forward(idx);
        torch::Tensor pos_emb = position_embedding_table->forward(torch::arange(T).to(device));
        torch::Tensor x = token_emb + pos_emb;
        for (auto& block : *blocks) {
            x = block->forward(x);
        }
        torch::Tensor logits = lm_head->forward(x);

        if (targets.defined()) {
            torch::Tensor loss = torch::nn::functional::cross_entropy(logits.view({-1, vocab_size}), targets.view({-1}));
            return logits, loss;
        }
        else {
            return logits, nullptr;
        }
    }

    torch::Tensor generate(torch::Tensor idx, int max_new_tokens) {
        for (int i = 0; i < max_new_tokens; ++i) {
            torch::Tensor idx_cond = idx.slice(1, -block_size, -1);
            torch::Tensor logits, _;
            std::tie(logits, _) = forward(idx_cond);
            logits = logits.select(1, -1);
            torch::Tensor probs = torch::softmax(logits, -1);
            torch::Tensor idx_next = torch::multinomial(probs, 1);
            idx = torch::cat({ idx, idx_next }, 1);
        }
        return idx;
    }

    torch::nn::Embedding token_embedding_table;
    torch::nn::Embedding position_embedding_table;
    torch::nn::Sequential blocks;
    torch::nn::LayerNorm ln_f;
    torch::nn::Linear lm_head;
};
TORCH_MODULE(BigramLanguageModel);

// Utility functions

std::vector<int64_t> read_data(const std::string& filename) {
    std::ifstream file(filename);
    std::vector<int64_t> data;
    std::string line;
    while (std::getline(file, line)) {
        for (char c : line) {
            data.push_back(stoi[c]);
        }
    }
    return data;
}

void initialize_vocabulary() { // pain
    chars = { ' ', '!', '"', '#', '$', '%', '&', '\'', '(', ')', '*', '+', ',', '-', '.', '/',
              '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?',
              '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
              'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', '^', '_',
              '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
              'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~' };

    vocab_size = chars.size();

    for (int i = 0; i < vocab_size; ++i) {
        stoi[chars[i]] = i;
        itos[i] = chars[i];
    }
}

int main() {
    initialize_vocabulary();

    train_data = read_data("train.txt");
    val_data = read_data("val.txt");

    torch::TensorOptions options(torch::kLong);
    torch::Tensor train_tensor = torch::from_blob(train_data.data(), { train_data.size() / block_size, block_size }, options).clone().to(device);
    torch::Tensor val_tensor = torch::from_blob(val_data.data(), { val_data.size() / block_size, block_size }, options).clone().to(device);

    torch::Device device = torch::cuda::is_available() ? torch::kCUDA : torch::kCPU;

    BigramLanguageModel model;
    model->to(device);

    torch::optim::Adam optimizer(model->parameters(), torch::optim::AdamOptions(learning_rate));

    int num_train_batches = train_tensor.size(0);
    int num_val_batches = val_tensor.size(0);

    torch::Tensor train_loss = torch::zeros({ max_iters });
    torch::Tensor val_loss = torch::zeros({ max_iters / eval_interval });

    std::default_random_engine generator;
    std::uniform_int_distribution<int> distribution(0, num_train_batches - 1);

    for (int iter = 0; iter < max_iters; ++iter) {
        model->train();

        int idx = distribution(generator);
        torch::Tensor batch = train_tensor.slice(0, idx, idx + batch_size).clone().to(device);

        optimizer.zero_grad();
        torch::Tensor _, loss = model->forward(batch, batch);
        loss.backward();
        optimizer.step();

        train_loss[iter] = loss.item<float>();

        if (iter % eval_interval == 0) {
            model->eval();

            torch::Tensor val_total_loss = torch::zeros({ eval_iters });
            for (int i = 0; i < eval_iters; ++i) {
                idx = distribution(generator);
                torch::Tensor val_batch = val_tensor.slice(0, idx, idx + batch_size).clone().to(device);

                torch::Tensor _, val_batch_loss = model->forward(val_batch, val_batch);
                val_total_loss[i] = val_batch_loss.item<float>();
            }

            float avg_val_loss = val_total_loss.mean().item<float>();
            val_loss[iter / eval_interval] = avg_val_loss;

            std::cout << "Iteration: " << iter << " | Train Loss: " << loss.item<float>() << " | Val Loss: " << avg_val_loss << std::endl;
        }
    }

    torch::save(model, "model.pt");

    return 0;
}

File: src/gpt_from_scratch/eda/v2.py
import torch
import torch.nn as nn
from torch.nn import functional as F

# hyperparameters
batch_size = 64  # how many independent sequences will we process in parallel
block_size = 256  # what is the maximum context length for predictions
max_iters = 5000  # how many training iterations
eval_interval = 500  # how often to evaluate our model on the validation set
learning_rate = 3e-4  # how fast should our model learn?
device = 'cuda' if torch.cuda.is_available() else 'cpu'  # is gpu available?
eval_iters = 200  # how many iterations to use when evaluating on the validation set
n_embd = 384
n_head = 6
n_layer = 6
dropout = 0.2

torch.manual_seed(1337)

# --- data ---
# wget https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt
with open("src/gpt_from_scratch/raw_data/input.txt", "r") as f:
    text = f.read()

# --- vocabulary ---
chars = sorted(list(set(text)))
vocab_size = len(chars)
# create a mapping from characters to integers (character level tokenizer)
stoi = {ch: i for i, ch in enumerate(chars)}
itos = {i: ch for i, ch in enumerate(chars)}
# encode a string s to a list of integers
def encode(s): return [stoi[c] for c in s]
# decode a list of integers l to a string
def decode(l): return ''.join([itos[i] for i in l])


# --- data preparation ---
data = torch.tensor(encode(text), dtype=torch.long, device=device)
n = int(0.9*len(data))  # first 90% of the data will be train, rest will be val
train_data = data[:n]
val_data = data[n:]

# --- model ---


def get_batch(split):
    # generate a small batch of data of inputs x and targets y
    data = train_data if split == 'train' else val_data
    ix = torch.randint(len(data) - block_size, (batch_size,))
    x = torch.stack([data[i:i+block_size] for i in ix])
    y = torch.stack([data[i+1:i+block_size+1] for i in ix])
    x, y = x.to(device), y.to(device)
    return x, y


@torch.no_grad()
def estimate_loss():
    out = {}
    model.eval()
    for split in ['train', 'val']:
        losses = torch.zeros(eval_iters)
        for k in range(eval_iters):
            X, Y = get_batch(split)
            logits, loss = model(X, Y)
            losses[k] = loss.item()
        out[split] = losses.mean()
    model.train()
    return out


class Head(nn.Module):
    """ one head of self-attention"""

    def __init__(self, head_size):
        super().__init__()
        self.key = nn.Linear(n_embd, head_size, bias=False)
        self.query = nn.Linear(n_embd, head_size, bias=False)
        self.value = nn.Linear(n_embd, head_size, bias=False)
        self.register_buffer('tril', torch.tril(
            torch.ones(block_size, block_size)))
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        B, T, C = x.shape
        k = self.key(x)  # (B, T, C)
        q = self.query(x)  # (B, T, C)
        # compute attention scores ("affinities")
        # (B, T, C) @ (B, T, C) -> (B, T, T)
        wei = q @ k.transpose(-2, -1) * C**-0.5
        wei = wei.masked_fill(
            self.tril[:T, :T] == 0, float('-inf'))  # (B, T, T)
        wei = F.softmax(wei, dim=-1)  # (B, T, T)
        wei = self.dropout(wei)
        # perform the weighted aggregation of the values
        v = self.value(x)  # (B, T, C)
        out = wei @ v  # (B, T, T) @ (B, T, C) -> (B, T, C)
        return out


class MultiHeadAttention(nn.Module):
    """ multiple heads of self-attention in parallel """

    def __init__(self, num_heads, head_size):
        super().__init__()
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        self.proj = nn.Linear(n_embd, n_embd)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        out = torch.cat([h(x) for h in self.heads], dim=-1)
        out = self.dropout(self.proj(out))
        return out


class FeedForward(nn.Module):
    """ a simple linear layer followed by a non-linearity """

    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, 4 * n_embd),
            nn.ReLU(),
            nn.Linear(4 * n_embd, n_embd),
            nn.Dropout(dropout)
        )

    def forward(self, x):
        return self.net(x)


class Block(nn.Module):
    """ Transformer block: communication followed by computation """

    def __init__(self, n_embd, n_head):
        # n_embd: embedding dimension, n_head: the number of heads we'd like
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedForward(n_embd)
        self.ln1 = nn.LayerNorm(n_embd)
        self.ln2 = nn.LayerNorm(n_embd)

    def forward(self, x):
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
        return x

# Simple Bigram Model


class BigramLanguageModel(nn.Module):

    def __init__(self):
        super().__init__()
        # each token directly reads off the logits for the next token from a lookup table
        self.token_embedding_table = nn.Embedding(vocab_size, n_embd)
        self.position_embedding_table = nn.Embedding(block_size, n_embd)
        self.blocks = nn.Sequential(
            *[Block(n_embd, n_head=n_head) for _ in range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embd)  # the final layer norm
        self.lm_head = nn.Linear(n_embd, vocab_size)

    def forward(self, idx, targets=None):
        B, T, = idx.shape

        # idx and targets are both (B, T) tensor of integers
        token_emb = self.token_embedding_table(idx)
        pos_emb = self.position_embedding_table(
            torch.arange(T, device=device))  # (T, C)
        x = token_emb + pos_emb  # (B, T, C)
        x = self.blocks(x)  # (B, T, C)
        logits = self.lm_head(x)  # (B, T, vocab_size)

        if targets is None:
            loss = None
        else:
            B, T, C = logits.shape
            # (B*T, C); this will conform to what torch expects
            logits = logits.view(B*T, C)
            targets = targets.view(B*T)
            # negative loss likelihood; which is cross entropy loss; measures the quality of the logits with respect to the targets
            loss = F.cross_entropy(logits, targets)

        # (B, T, C); which is Batch (4), Time (8), and Channel (65) tensors of integers, and is the score, or next character in a sequence
        return logits, loss

    def generate(self, idx, max_new_tokens):
        # idx is (B, T) array of indices in the current context
        for _ in range(max_new_tokens):
            # crop idx to the last block_size tokens
            idx_cond = idx[:, -block_size:]
            # get the prediction
            logits, loss = self(idx_cond)
            # focus only on the last time step
            logits = logits[:, -1, :]  # becomes (B, C) tensor
            # apply softmax to convert to probabilities
            probs = F.softmax(logits, dim=-1)  # (B, C) tensor
            # sample from the distribution
            idx_next = torch.multinomial(probs, num_samples=1)  # (B, 1) tensor
            # append sampled index to the running sequence
            idx = torch.cat((idx, idx_next), dim=1)  # (B, T+1) tensor
        return idx


model = BigramLanguageModel()
m = model.to(device)
print(sum(p.numel() for p in m.parameters()), 'M parameters')

# creating the pytorch optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)

for iter in range(max_iters):

    # every once in a while we will evaluate our model on the validation set
    if iter % eval_interval == 0:
        losses = estimate_loss()
        print(
            f"Step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the lsos
    logits, loss = model(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

# --- generation ---
context = torch.zeros((1, 1), dtype=torch.long,
                      device=device)  # (B=1, T=1) tensor
print(decode(m.generate(context, max_new_tokens=500)[0].tolist()))
# open('decoder_transformer_output.txt', 'w').write(decode(m.generate(context(max_new_tokens=10000)[0].tolist())))
